README.md
# GSEAbar
Barplot for GSEA result


NEWS.md
## 0.1.0

### 定义基本函数

-   **gseabar()** 按p值颜色排序
-   gseabar2() 按p值大小排序

## 0.1.1

修改部分参数

## 0.1.2

-   revers参数添加了排序方式

-   divide参数新增双侧分别显示通路名称 （文本过长时不显示，后期想办法解决）


R/gseabar2.R
##' @title gseabar2
##' @name gseabar2
##' @param object GSEA enrich results.
##' @param n number of categories to show
##' @param font.size font size
##' @param title plot title
##' @param length a numeric value sets wrap length, alternatively a
##' custom function to format axis labels.
##' by default wraps names longer that 40 characters
##' @param color manual colores for bar plot
##' @param reverse order
##' @param divide divide the label by group
##' @param ... other parameter, ignored

##' @import ggplot2
##' @import DOSE
##' @import stringr
##' @import dplyr
##' @import ggfun
##' @import forcats
##' @import RColorBrewer
##' @return ggplot2 object
##' @export
##'
##' @examples
##' \donttest{
##' # load data
##' test_data <- system.file("extdata", "gseares.RDS", package = "GSEAbar")
##' gseares <- readRDS(test_data)
##'
##' # bar plot for GSEA
##' gseabar2(gseares,
##'        n=12,
##'        font.size=12,
##'        title="GSEA-barplot",
##'        length=40,
##'        color = c("#f87669","#2874C5"),
##'        reverse= FALSE,
##'        divide=FALSE)
##' }


# define function

gseabar2 <- function(object,
                    n=12,
                    font.size=10,
                    title="",
                    length=40,
                    color = c("#f87669","#2874C5"),
                    reverse= FALSE,
                    divide=FALSE,...) {
  ## use *gsdata* to satisy barplot generic definition
  ## actually here is an gseaResult object.

  gsdata <- data.frame(object)
  #### add a new sign group by |NES|>0
  gsdata$sign<-ifelse(gsdata$NES>0,"Activated","Suppressed")
  #### add a new group which defined the -log10 pvalue negative or positive
  gsdata$pl = ifelse(gsdata$NES>0,-log10(gsdata$p.adjust),
                     log10(gsdata$p.adjust))

   p <- gsdata %>%
    group_by(sign) %>%
    arrange(p.adjust) %>%
    slice(1:n) %>%
    ggplot( aes(pl, fct_reorder(Description, pl),fill=sign))+theme_bw(font.size)
### 是否更换排序方式
   if (reverse) {
     p <- gsdata %>%
       group_by(sign) %>%
       arrange(p.adjust) %>%
       slice(1:n) %>%
       ggplot( aes(pl, fct_reorder(Description, desc(pl)),fill=sign))+theme_bw(font.size)
   }
### 是否分侧显示结果名称
   if (divide){
     plotdata<-gsdata %>%
       group_by(sign) %>%
       arrange(p.adjust) %>%
       slice(1:n)

     p<-p+theme(axis.text.y = element_blank(),
                 axis.ticks = element_blank(),
                 panel.border = element_blank())+
       geom_text(aes(x=ifelse(plotdata$NES>1,-0.1,0.1),label=plotdata$Description),
                 hjust=ifelse(plotdata$NES>1,1,0))
   }
### 定义文本长度
  label_func <- default_labeller(length)
  if(is.function(length)) {
    label_func <- length
  }

  p + geom_col(orientation='y') + # geom_bar(stat = "identity") + coord_flip() +
    scale_fill_manual(values=color)+
    scale_y_discrete(labels = label_func) +
    ggtitle(title) + ylab(NULL) +  xlab("-log10(pval)")+
    theme(plot.title = element_text(hjust = 0.5)) +
    theme(legend.background=element_roundrect(color="#808080", linetype=4))
}


R/utillties.R
##' wraping long string to multiple lines
##'
##'
##' @title str_wrap
##' @param string input string
##' @param width the maximum number of characters before wrapping to a new line
##' @return update strings with new line character inserted
##' @export
##' @author Guangchuang Yu and Erqiang Hu
str_wrap <- function(string, width = getOption("width")) {
  # x <- gregexpr(' ', string)
  # vapply(seq_along(x),
  #        FUN = function(i) {
  #            y <- x[[i]]
  #            n <- nchar(string[i])
  #            len <- (c(y,n) - c(0, y)) ## length + 1
  #            idx <- len > width
  #            j <- which(!idx)
  #            if (length(j) && max(j) == length(len)) {
  #                j <- j[-length(j)]
  #            }
  #            if (length(j)) {
  #                idx[j] <- len[j] + len[j+1] > width
  #            }
  #            idx <- idx[-length(idx)] ## length - 1
  #            start <- c(1, y[idx] + 1)
  #            end <- c(y[idx] - 1, n)
  #            words <- substring(string[i], start, end)
  #            paste0(words, collapse="\n")
  #        },
  #        FUN.VALUE = character(1)
  # )
  result <- vapply(string,
                   FUN = function(st) {
                     words <- list()
                     i <- 1
                     while(nchar(st) > width) {
                       if (length(grep(" ", st)) == 0) break
                       y <- gregexpr(' ', st)[[1]]
                       n <- nchar(st)
                       y <- c(y,n)
                       idx <- which(y < width)
                       # When the length of first word > width
                       if (length(idx) == 0) idx <- 1
                       # Split the string into two pieces
                       # The length of first piece is small than width
                       words[[i]] <- substring(st, 1, y[idx[length(idx)]] - 1)
                       st <- substring(st, y[idx[length(idx)]] + 1, n)
                       i <- i + 1
                     }
                     words[[i]] <- st
                     paste0(unlist(words), collapse="\n")
                   },
                   FUN.VALUE = character(1)
  )
  names(result) <- NULL
  result
}

##' Detect the presence or absence of a pattern at the beginning or end of a string or string vector.
##'
##'
##' @title str_starts
##' @rdname str-starts-ends
##' @param string input string
##' @param pattern pattern with which the string starts or ends
##' @param negate if TRUE, return non-matching elements
##' @return a logical vector
##' @export
##' @author Guangchuang Yu
str_starts <- function(string, pattern, negate=FALSE) {
  pattern <- paste0('^', pattern)
  str_detect(string, pattern, negate)
}

##' @rdname str-starts-ends
##' @export
str_ends <- function(string, pattern, negate=FALSE) {
  pattern <- paste0(pattern, '$')
  str_detect(string, pattern, negate)
}

##' @importFrom stats setNames
str_detect <- function(string, pattern, negate) {
  res <- setNames(
    vapply(string, grepl, pattern=pattern,
           FUN.VALUE=logical(1)),
    NULL)
  if (negate) res <- !res
  return(res)
}



##' Get the distance of the label
##'
##' @param dimension one of 1 and 2
##' @param label_location label_location
##' @noRd
get_label_diss <- function(dimension, label_location) {
  nn <- nrow(label_location)
  label_dis <- matrix(NA, nrow = nn, ncol = nn)
  colnames(label_dis) <- rownames(label_dis) <- label_location$label
  for (i in seq_len(nn - 1)) {
    for (j in (i + 1):nn) {
      label_dis[i ,j] <- label_location[i, dimension] -  label_location[j, dimension]
    }
  }
  label_diss <- reshape2::melt(label_dis)
  label_diss <- label_diss[label_diss[,1] != label_diss[,2], ]
  label_diss <- label_diss[!is.na(label_diss[,3]), ]
  label_diss[, 1] <- as.character(label_diss[, 1])
  label_diss[, 2] <- as.character(label_diss[, 2])
  return(label_diss)
}



# adjust_location <- function(label_location, x_adjust, y_adjust) {
# label_diss_x <- get_label_diss(1, label_location)
# label_diss_y <- get_label_diss(2, label_location)

# label_diss_large <- which(abs(label_diss_y[, 3]) < y_adjust) %>%
# intersect(which(label_diss_y[, 3] > 0)) %>%
# intersect(which(abs(label_diss_x[, 3]) < x_adjust))

# label_diss_small <- which(abs(label_diss_y[, 3]) < y_adjust) %>%
# intersect(which(label_diss_y[, 3] < 0)) %>%
# intersect(which(abs(label_diss_x[, 3]) < x_adjust))

# label_location[label_diss_y[label_diss_large, 1], 2] <- label_location[label_diss_y[label_diss_large, 2], 2] + y_adjust
# label_location[label_diss_y[label_diss_small, 1], 2] <- label_location[label_diss_y[label_diss_small, 2], 2] - y_adjust
# return(label_location)
# }



#' default_labeller
#'
#' default labeling function that uses the
#' internal string wrapping function `GSEAbar::str_wrap`
#' @noRd
default_labeller <- function(n) {
  function(str){
    str <- gsub("_", " ", str)
    str_wrap(str, n)
  }
}




R/gseabar.R
##' @title gseabar
##' @name gseabar
##' @param object GSEA enrich results.
##' @param color one of 'pvalue', 'p.adjust', 'qvalue', 'qvalues', 'pavl' or 'padj'
##' @param n number of categories to show
##' @param font.size font size
##' @param title plot title
##' @param reverse order
##' @param divide divide the label by group
##' @param length a numeric value sets wrap length, alternatively a
##' custom function to format axis labels.
##' by default wraps names longer that 40 characters
##' @param ... other parameter, ignored

##' @import ggplot2
##' @import DOSE
##' @import stringr
##' @import dplyr
##' @import ggfun
##' @import forcats
##' @import RColorBrewer
##' @return ggplot2 object
##' @export
##'
##' @examples
##' \donttest{
##' # load data
##' test_data <- system.file("extdata", "gseares.RDS", package = "GSEAbar")
##' gseares <- readRDS(test_data)
##'
##' # bar plot for GSEA
##' gseabar(gseares,
##'        color='p.adjust',
##'        n=12,
##'        font.size=12,
##'        title="GSEA-barplot",
##'        length=40,
##'        reverse= FALSE,
##'        divide=FALSE)
##' }

# define function

  gseabar <- function(object,
                      color="p.adjust",
                      n=10,
                      font.size=12,
                      title="",
                      length=40,
                      reverse= FALSE,
                      divide = FALSE,...) {
    ## use *gsdata* to satisy barplot generic definition
    ## actually here is an gseaResult object.

    gsdata <- data.frame(object)

    colorBy <- match.arg(color, c("pvalue", "p.adjust", "qvalue", "qvalues","pval","padj"))
  #### add a new sign group by |NES|>0
  # gsdata$sign<-ifelse(gsdata$NES>0,"Activated","Suppressed")
      p <- gsdata %>%
        group_by(sign(NES)) %>%
        arrange(pvalue) %>%
        slice(1:n) %>%
        ggplot(aes(NES, fct_reorder(Description, NES)))+
        theme_bw(font.size)

### 是否更换排序方式
    if (reverse){
      p <- gsdata %>%
        group_by(sign(NES)) %>%
        arrange(p.adjust) %>%
        slice(1:n) %>%
        ggplot(aes(NES, fct_reorder(Description, desc(NES))))+
        theme_bw(font.size)
    }
### 是否分侧显示结果名称
    if (divide){
      plotdata<-gsdata %>%
        group_by(sign(NES)) %>%
        arrange(p.adjust) %>%
        slice(1:n)

      p<-p+theme(axis.text.y = element_blank(),
                 axis.ticks = element_blank(),
                 panel.border = element_blank())+
        geom_text(aes(x=ifelse(plotdata$NES>1,-0.05,0.05),label=plotdata$Description),
                  hjust=ifelse(plotdata$NES>1,1,0))
    }

### 定义文本长度
      label_func <- default_labeller(length)
      if(is.function(length)) {
        label_func <- length
      }

    p +
      scale_fill_continuous(low="#f87669", high="#2874C5",
                            guide=guide_colorbar(reverse=TRUE))+
      geom_col(aes_string(fill=colorBy)) + # geom_bar(stat = "identity") + coord_flip() +
      scale_y_discrete(labels = label_func) +
      ggtitle(title) + ylab(NULL) +  xlab("(Suppressed) ---NES--- (Activated)")+
      theme(plot.title = element_text(hjust = 0.5)) +
      theme(legend.background=element_roundrect(color="#808080", linetype=4))

  }

